### PR TITLE
feat: "the merge" rpc scaffolding

### LINKED ISSUE #1248: Eth1 <-> Eth2: First RPC Connection PoC
For quite some time now we are always side-tracked with some other fancy stuff like the `EIP-1559` implementation and have "The Merge" always laying around on the sideline without having the time for some progress.

We should break through this mental barrier by just give this some descent start! 😜

A good first task here would be to give it a start by looking at the [Rayonism](https://rayonism.io/) hackathon results and specs and implement one of the RPC consensus calls from the [spec](https://github.com/ethereum/rayonism/blob/master/specs/merge.md#consensus-json-rpc), then fire-up one of the prepared [Eth 2.0 clients](https://notes.ethereum.org/@protolambda/rayonism#Eth2-Clients) and see if we can get a connection (this should actually get some proof in form of a screenshot of the console output or something on implementation 😁 ). Client RPC call implementation can be found [here](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/client/lib/rpc).

This is *just* for having some first infrastructure setup ready as some initial starting point, so this doesn't even implies to implement the functionality itself from the respective RPC call. So it is enough to proof that the two clients have tried to start talking with each other. The RPC functionality implementation can then be done in a subsequent second step as a separate PR.

### PR DESCRIPTION
**Pull request by gabrocheleau**
This PR creates the basic structure to begin working on "The Merge" by implementing placeholder methods for [these 4 JSON-RPC methods](https://github.com/ethereum/rayonism/blob/master/specs/merge.md#consensus-json-rpc) and adding basic validation tests for each method. (Resolves #1248)

Here's the procedure I used to establish a connection between an ETH 2.0 "merge" client and an ethereumjs client:

1. Clone this Teku repo: https://github.com/txrx-research/teku/tree/rayonism-merge 
2. Install the Teku (ETH 2.0) client: `./gradlew distTar installDist`
3. Install and run the ethereumjs client with RPC server & no syncing: `npm run client:start --rpc --maxPeers=0`. 
4. Run the Teku client, connecting to the ethereumjs ETH 1.0 client: `./build/install/teku/bin/teku ----eth1-endpoint=http://localhost:8545`
5. The ETH 2.0 client will now attempt to connect to the ETH 1.0 client. 

As @holgerd77 found out we can view the JSON-RPC requests by manually logging all requests by adding the following in the `runRPCServer` function located in `packages\client\bin\cli.ts` (after `server.http().listen(rpcport)`): 
```
  server.on('request', function (req) {
    console.log('Request')
    console.log(req)
  })
```
Hope that's helpful! 😀

### CODE REVIEW

**Comment by holgerd77** [comment]
Very cool! 😄  👍

File: packages/client/lib/rpc/modules/consensus.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import { middleware, validators } from '../validation'
+import type { Chain } from '../../blockchain'
+import type { EthereumClient } from '../..'
+import type { EthereumService } from '../../service'
+import type { EthProtocol } from '../../net/protocol'
+import type VM from '@ethereumjs/vm'
+
+/**
+ * eth_* RPC module
```
Comment (ryanio) [inline packages/client/lib/rpc/modules/consensus.ts:9]:
```suggestion
 * consensus_* RPC module
```

File: packages/client/lib/rpc/modules/consensus.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ * @memberof module:rpc/modules
+ */
+export class Consensus {
+  private _chain: Chain
+  private _vm: VM | undefined
+  public ethVersion: number
+
+  /**
+   * Create eth_* RPC module
```
Comment (ryanio) [inline packages/client/lib/rpc/modules/consensus.ts:18]:
```suggestion
   * Create consensus_* RPC module
```

File: packages/client/lib/rpc/modules/consensus.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+   * @param client Client to which the module binds
+   */
+  constructor(client: EthereumClient) {
+    const service = client.services.find((s) => s.name === 'eth') as EthereumService
+    this._chain = service.chain
+    this._vm = (service.synchronizer as any)?.execution?.vm
+
+    const ethProtocol = service.protocols.find((p) => p.name === 'eth') as EthProtocol
+    this.ethVersion = Math.max(...ethProtocol.versions)
```
Comment (ryanio) [inline packages/client/lib/rpc/modules/consensus.ts:27]: you won't need this since it was just for `eth_protocolVersion`
```suggestion
```

Comment (gabrocheleau) [reply packages/client/lib/rpc/modules/consensus.ts:27]: Thanks resolved. Also removed the `ethVersion` property from the class declaration

**Comment by holgerd77** [comment]
Pretty cool! Tests are still failing atm (some linting, run `npm run lint:fix` from the client folder, eventually needs manual fixes). 

Also: can we add some (very basic I guess) tests here along which just somewhat dummy-call into the methods? Then we have this complete here and coverage is not falling so harsh on the PR.

**Comment by gabrocheleau** [comment]
> Pretty cool! Tests are still failing atm (some linting, run `npm run lint:fix` from the client folder, eventually needs manual fixes).
> 
> Also: can we add some (very basic I guess) tests here along which just somewhat dummy-call into the methods? Then we have this complete here and coverage is not falling so harsh on the PR.

Sure! Will write a few tests today.

File: packages/client/lib/rpc/modules/consensus.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+   *   * stateRoot - the root of the final state trie of the block.
+   *   * number - the block number.
+   *   * gasLimit - the maximum gas allowed in this block.
+   *   * gasUsed - the total used gas by all transactions in this block.
+   *   * timestamp - unix timestamp of the block; must be equal to the value of timestamp parameter.
+   *   * receiptsRoot - the root of the receipts trie of the block.
+   *   * logsBloom - the bloom filter for the logs of the block.
+   *   * transactions - Array of encoded transactions, each transaction is a byte list, representing TransactionType || TransactionPayload or LegacyTransaction as defined in EIP 2718
+   */
+  async assembleBlock(params: [string, number]) {
```
Comment (holgerd77) [inline packages/client/lib/rpc/modules/consensus.ts:50]: To have these extensive code docs already in will actually be super helpful on the actual implementation, thanks for taking the time here! 😄 

(@ryanio this one would likely be a case for the block builder? 🙂  👷)

**Review by holgerd77 — APPROVED** [review]
Looks good, first Eth 2.0 PR to be merged, cool! 🥳

**Pull request merged**